control-plane/cni/main.go

```diff
@@ -0,0 +1,223 @@
+	K8S_POD_INFRA_CONTAINER_ID types.UnmarshallableString
+}
+
+// PluginConf is whatever you expect your configuration json to be. This is whatever
+// is passed in on stdin. Your plugin may wish to expose its functionality via
+// runtime args, see CONVENTIONS.md in the CNI spec.
+type PluginConf struct {
+	// This embeds the standard NetConf structure which allows your plugin
+	// to more easily parse standard fields like Name, Type, CNIVersion,
+	// and PrevResult.
+	types.NetConf
+
+	RuntimeConfig *struct {
+		SampleConfig map[string]interface{} `json:"sample_config"`
+	} `json:"runtime_config"`
+
+	Name string `json:"name"`
+	// Type of plugin (consul-cni)
+	Type string `json:"type"`
+	// CNIBinDir is the location of the cni config files on the node. Can bet as a cli flag.
+	CNIBinDir string `json:"cni_bin_dir"`
+	// CNINetDir is the locaion of the cni plugin on the node. Can be set as a cli flag.
+	CNINetDir string `json:"cni_net_dir"`
+	// Multus is if the plugin is a multus plugin. Can be set as a cli flag.
+	Multus bool `json:"multus"`
+	// Kubeconfig file name. Can be set as a cli flag.
+	Kubeconfig string `json:"kubeconfig"`
+	// LogLevl is the logging level. Can be set as a cli flag.
+	LogLevel string `json:"log_level"`
+}
+
+// parseConfig parses the supplied configuration (and prevResult) from stdin.
+func parseConfig(stdin []byte) (*PluginConf, error) {
+	cfg := PluginConf{}
+
+	if err := json.Unmarshal(stdin, &cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse network configuration: %v", err)
+	}
+
+	// Parse previous result. This will parse, validate, and place the
+	// previous result object into conf.PrevResult. If you need to modify
+	// or inspect the PrevResult you will need to convert it to a concrete
+	// versioned Result struct.
+	if err := version.ParsePrevResult(&cfg.NetConf); err != nil {
+		return nil, fmt.Errorf("could not parse prevResult: %v", err)
+	}
+	// End previous result parsing
+
+	// Do any validation here
+	// TODO: Do validation
+	//	if conf.AnotherAwesomeArg == "" {
+	//		return nil, fmt.Errorf("anotherAwesomeArg must be specified")
+	//	}
+
+	return &cfg, nil
+}
+
+// cmdAdd is called for ADD requests.
+func cmdAdd(args *skel.CmdArgs) error {
+	cfg, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	// Get the values of args passed through CNI_ARGS
+	cniArgs := CNIArgs{}
+	if err := types.LoadArgs(args.Args, &cniArgs); err != nil {
+		return err
+	}
+
+	podNamespace := string(cniArgs.K8S_POD_NAMESPACE)
+	podName := string(cniArgs.K8S_POD_NAME)
+
+	// We only run in a pod
+	if podNamespace == "" && podName == "" {
+		return fmt.Errorf("not running in a pod, namespace and pod should have values")
+	}
+
+	logPrefix := fmt.Sprintf("%s/%s", podNamespace, podName)
+	logger := hclog.New(&hclog.LoggerOptions{
+		Name:  logPrefix,
+		Level: hclog.LevelFromString(cfg.LogLevel),
+	})
+
+	// Check to see if the plugin is a chained plugin
+	if cfg.PrevResult == nil {
```

Comment:
Chained plugins pass along their result from their runs. We are not doing anything that needs passing on so we just forward on the previous result. We are designed to be the final plugin in the chain but 🤷. 

Tidbit: Kindnet, KinD's CNI plugin, is not a proper plugin and it does not pass along its previous result. This is why we can not run our consul-cni plugin directly on KinD... it would just throw errors. It took me a while to figure this out as I thought for a while that the problem was with my plugin... For KinD we need to disable Kindnet and install another plugin before ours, like Calico.

Reply:
I added 'final' in the message.